File: frontend/types/resolution.ts
export type ResolutionStatus = "not_started" | "in_progress" | "completed"
export type AISignal = "on_track" | "at_risk" | "no_signal"

export interface AIPlanMonth {
  month_label: string
  goal: string
  actions: string[]
}

export interface ProgressLogResponse {
  id: string
  progress_percent: number
  note: string | null
  logged_at: string
}

export interface CheckInResponse {
  id: string
  resolution_id: string
  year: number
  month: number
  rating: number
  note: string | null
  created_at: string
  updated_at: string
}

export interface ResolutionResponse {
  id: string
  title: string
  description: string | null
  status: ResolutionStatus
  progress_percent: number
  target_date: string | null
  ai_plan: AIPlanMonth[] | null
  created_at: string
  updated_at: string
  check_ins: CheckInResponse[]
  progress_logs: ProgressLogResponse[]
}

export interface CheckInCreateInput {
  year: number
  month: number
  rating: number
  note?: string
}

export interface ResolutionCreateInput {
  title: string
  description?: string
  target_date?: string
}

export interface ResolutionUpdateInput {
  title?: string
  description?: string
  status?: ResolutionStatus
  progress_percent?: number
  target_date?: string
}

export interface AIAnalysisItem {
  resolution_id: string
  resolution_title: string
  signal: AISignal
  evidence: string[]
  suggestion: string
}

export interface AIAnalysisResponse {
  generated_at: string
  analyses: AIAnalysisItem[]
}

File: backend/app/schemas/resolution.py
from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator

VALID_STATUSES = ["not_started", "in_progress", "completed"]


class AIPlanMonth(BaseModel):
    month_label: str
    goal: str
    actions: list[str]

    model_config = ConfigDict(from_attributes=True)


class ResolutionCreate(BaseModel):
    title: str = Field(min_length=1)
    description: str | None = None
    target_date: date | None = None


class ResolutionUpdate(BaseModel):
    title: str | None = None
    description: str | None = None
    status: str | None = None
    progress_percent: int | None = Field(None, ge=0, le=100)
    target_date: date | None = None
    ai_plan: list[dict] | None = None

    @field_validator("status")
    @classmethod
    def validate_status(cls, v: str | None) -> str | None:
        if v is not None and v not in VALID_STATUSES:
            raise ValueError(f"status must be one of {VALID_STATUSES}")
        return v


class ProgressLogCreate(BaseModel):
    progress_percent: int = Field(ge=0, le=100)
    note: str | None = None


class ProgressLogResponse(BaseModel):
    id: UUID
    progress_percent: int
    note: str | None
    logged_at: datetime

    model_config = ConfigDict(from_attributes=True)


class CheckInCreate(BaseModel):
    year: int
    month: int
    rating: int = Field(ge=1, le=5)
    note: str | None = None


class CheckInResponse(BaseModel):
    id: UUID
    resolution_id: UUID
    year: int
    month: int
    rating: int
    note: str | None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ResolutionResponse(BaseModel):
    id: UUID
    title: str
    description: str | None
    status: str
    progress_percent: int
    target_date: date | None
    ai_plan: list[AIPlanMonth] | None = None
    created_at: datetime
    updated_at: datetime
    check_ins: list[CheckInResponse] = []
    progress_logs: list[ProgressLogResponse] = []

    model_config = ConfigDict(from_attributes=True)


class AIAnalysisItem(BaseModel):
    resolution_id: str
    resolution_title: str
    signal: str  # "on_track" | "at_risk" | "no_signal"
    evidence: list[str]
    suggestion: str


class AIAnalysisResponse(BaseModel):
    generated_at: str  # ISO date
    analyses: list[AIAnalysisItem]
